.github/workflows/ci.yml

```diff
@@ -35,43 +20,79 @@ jobs:
       matrix:
         include:
           - name: Linux
+            id: linux
             cache_id: linux
             os: ubuntu-22.04-16core
             flags: ""
           - name: Linux Nightly
+            id: linux-nightly
             cache_id: linux
             os: ubuntu-22.04-16core
             flags: "--features nightly,test_features"
           - name: MacOS
+            id: macos
             cache_id: macos
             os: macos-latest-xlarge
             # FIXME: some of these tests don't work very well on MacOS at the moment. Should fix
             # them at earliest convenience :)
-            flags: "--exclude integration-tests --exclude node-runtime --exclude runtime-params-estimator --exclude near-network --exclude estimator-warehouse"
+            # Note that integration-tests are disabled at the step running them in the job below.
+            flags: "--exclude node-runtime --exclude runtime-params-estimator --exclude near-network --exclude estimator-warehouse"
     timeout-minutes: 90
     steps:
       # Some of the tests allocate really sparse maps, so heuristic-based overcommit limits are not
       # appropriate here.
       # FIXME(#9634): remove this once the issue is resolved.
       - run: sudo sysctl vm.overcommit_memory=1 || true
       - uses: actions/checkout@v4
+
+      # Install all the required tools
       - uses: baptiste0928/cargo-install@21a18ba3bf4a184d1804e8b759930d3471b1c941
         with:
           crate: cargo-nextest
       - uses: baptiste0928/cargo-install@21a18ba3bf4a184d1804e8b759930d3471b1c941
         with:
           crate: cargo-deny
+      - uses: baptiste0928/cargo-install@21a18ba3bf4a184d1804e8b759930d3471b1c941
+        with:
+          crate: cargo-llvm-cov
+
+      # Setup the dependency rust cache and llvm-cov
       - uses: Swatinem/rust-cache@a95ba195448af2da9b00fb742d14ffaaf3c21f43
         with:
           prefix-key: "0" # change this to invalidate CI cache
           shared-key: "cargo_nextest-${{ matrix.cache_id }}"
-      - run: cargo nextest run --locked --workspace -p '*' --cargo-profile quick-release --profile ci ${{ matrix.flags }}
+      - run: cargo llvm-cov show-env | tr -d "'" >> "$GITHUB_ENV"
+
+      # Run unit tests
+      - run: cargo nextest run --locked --workspace --exclude integration-tests --cargo-profile quick-release --profile ci ${{ matrix.flags }}
         env:
           RUST_BACKTRACE: short
+      - run: cargo llvm-cov report --profile quick-release --codecov --output-path unittests.json
+      - uses: codecov/codecov-action@eaaf4bedf32dbdc6b720b63067d99c4d77d6047d
+        with:
+          files: unittests.json
+          fail_ci_if_error: true
+          flags: unittests,${{ matrix.id }}
+      # See https://github.com/taiki-e/cargo-llvm-cov/issues/292
+      - run: find target -name '*.profraw' -delete
+
+      # Run integration tests
```

Comment:
TBH I think #9415 should be solved by using something like `just test`, not by having everything stuffed into nextest, among other reasons because if we did it we could move `clippy` out and have a cache for it, not needing to rebuild it from scratch each time :)

Also, this split doesn’t change the fact that `cargo nextest run` still runs all the tests and works fine. I do think the categorization within the tool is useful, especially considering yesterday’s discussion that debugging integration tests is very bad dev UX, and thus we’d mostly be interested in the code coverage for unit-tests only.

Reply:
Well, so back in buildkite days we also had "cargo nextest run", but there were like 5 different ways to run integration tests and you had to pay really close attention to which specific invocation was failing – one of them failing would usually mean the others were passing.

This was exactly what prompted me to start working on CI in the first place. My honest evaluation is that we got into that state back then precisely *because* it was normal and natural to "just add another invocation of `cargo test/nextest/etc` with my snowflake combination of flags".

Having a single `cargo nextest` invocation I believe does a great job pushing against such configurations (and I hope it will eventually push enough to obviate the need for nightly vs. non-nightly configurations as well.) Splitting invocations into parts goes back towards normalizing having more of the snowflake configurations.

Now, to make this actionable at all – I wonder if nextest could be made to export test coverage on a per-test basis. All of the tests are distinct binary invocations, which should allow us to produce coverage files for each individual test (and therefore could then eventually be filtered by e.g. crate, module, etc. *after the fact*), not just two broad categories of unit vs integration tests.

Unfortunately `cargo llvm-cov nextest` appears to just run `cargo nextest` after setting up the environment, but *perhaps* [target runner scripts](https://nexte.st/book/target-runners.html) could be used as a mechanism to isolate and produce coverage information for each individual step?

Reply:
> we could move clippy out and have a cache for it

I don’t see why we couldn’t have a cache right now. We just need to *also* cache the `target/style` directory…

FWIW my belief is that the current setup with `target/style` is actually better for local development purposes as developers no longer risk accidentally invalidating their cached proper build when they run `cargo clippy -p foo` or something of the sort (by the virtue of not needing to run `cargo clippy` manually at all.) Though now that I write this, I realized there’s a bug in how style tests set the `TARGET_DIR` – it should be different for each test that builds code (in particular `themis` should have its own `TARGET_DIR` so it does not get blocked waiting on the target directory lock…)

Reply:
> (does rust-cache not do it…?)

Considering the clippy test is always among the last ones to complete and is written as slow (so 120+ seconds), I’d guess it doesn’t do it.

Counter-proposal: what if we had a Justfile with:
- A rule for unit tests
- A rule for clippy that runs it with target_dir set to `target/style`
- A rule for integration tests
- A `just test` rule that runs all of them

Then developers could run `just test`, and CI could run the rules independently, and maybe even override the target folder depending on whether we can fix target folder caching some other way. WDYT?

(I think changes to llvm-cov or cargo nextest might be a great idea as there are already many profraw files and we’d "just" need to figure out which are the right ones; but will not be reasonably doable before november 26, especially as I’m OOO the second half of next week and have a lot of other stuff to do by then)

Reply:
Pulling in node (just-test is a javascript thing, isn’t it?) just for development workflow is probably not all that great, but if there’s a similar Rust or Python tool, I wouldn’t object _too_ much… Would it really be that onerous to write out own Rust CLI tool, just like we have themis?

One reason why I think making our own tool might be important is to not lock ourselves into the interface that some off-shelf tool provides here. Like, for example, I personally frequently run subset of the test suite locally (e.g. `cargo nextest -p 'near-vm*'`) and I think it is important that this gets preserved (without making it necessary for people to *think* about whether they’re running `nextest` or some other tool.)

Reply:
> (I think changes to llvm-cov or cargo nextest might be a great idea as there are already many profraw files and we’d "just" need to figure out which are the right ones; but will not be reasonably doable before november 26, especially as I’m OOO the second half of next week and have a lot of other stuff to do by then)

The runner option wouldn’t really require any changes to either of the tools, but rather just a different way to invoke them. I think I could dedicate a few days to prototype this and we can set a deadline to land this before your time off if we don’t find any time to come up with a better solution.

But I’m also not quite sure why do we need this deadline in the first place. Although lack of coverage metrics was pointed out in a review, it isn’t like landing this PR before November 26th will fundamentally change much, will it? After all, we’ve already have seen some preliminary results up above and *some* conclusions can already be made even now if one looks at the [current absolute coverage numbers](https://app.codecov.io/gh/near/nearcore/pull/10180/tree).

Reply:
> (just-test is a javascript thing, isn’t it?)

I’m thinking of [just](https://github.com/casey/just), which is a rust simplified version of make, basically, focusing on command running rather than building :)

Considering this, WDYT about landing this, and me taking as an action item to add a `Justfile` for running tests to our repo as soon as I get some time? (it’ll also allow us to use the single `just test` command to also run the python tests, which are currently outside nextest)

> it isn’t like landing this PR before November 26th will fundamentally change much, will it?

It’ll allow us to have the report write "they fixed it" for this action item at least, which is better for PR than "they’re still working on it" :)

Reply:
Actually, relatedly, in order to be able to use the github actions for eg. lychee (that I’m introducing in https://github.com/near/nearcore/pull/10217), we’ll need to run the tests outside of cargo nextest anyway, unless we want to check them twice in CI. In particular, we could start using [the clippy action](https://github.com/marketplace/actions/clippy-action), which AFAIU has nicer error reporting than our current "manually read the logs" system.

So I think this confirms the want for a tool like `just`, so that we could have `just test` run `lychee` in a way similar to what is being run by CI :)

Reply:
Hm, so the weekend gave me ample of time to consider the last message. I agree that having nice reporting within GitHub is really nice.

But I also see a major (I think) flaw in what I think you’re proposing we should do. To double check, are you suggesting that we maintain `.github/workflows/ci.yaml` *and* a `Justfile` *and* that they would each implement the same checks independently? e.g. where `Justfile` would run `cargo clippy`, the GHA would run the clippy action? I think doing so is precisely the mistake I would like to avoid. If we’re adding a Justfile, the CI could run exactly the `just` commands that ought to be run as part of a local development flow (and ideally that command is singular or is split up into sub-jobs programatically) but that does not come with the nice error reporting benefits.

But unless we eschew those benefits, I don’t see how can we ensure, especially in the long term, that the checks run via the justfile remain synchronized with those run within GHA. And that means that ultimately we would still sometimes see instances where a successful `just` pass does not guarantee the CI passing (or vice versa, for that matter.)

Reply:
My current goal is indeed to have the CI run `just nextest-unit && coverage && just nextest-integration && coverage`, where `just nextest` for local use would also run the two.

However, for `clippy`, in order to get the nice reporting, I would indeed have a `just clippy` for local development that would not be DRY with the `clippy-action` for github, because by definition `clippy-action` cannot know to use our `Justfile` (or literally any other way to run all the tests we could have), any more than `just` could automatically figure out the command `clippy-action` runs.

So I’d say that by definition, I’d say that we can only have two of the following three:
- nice error reporting
- not implementing error reporting ourselves
- locally-run command is DRY with remotely-run command

I also personally think we should choose the first two, and whenever we come across an instance where `just` and the CI workflow don’t give the same result we can consider that as a bug and treat it by probably just copy-pasting again from the updated upstream `clippy-action`.

That being said, this PR is only about `nextest` splitting, so I’d argue that the discussion about clippy lints could be postponed for the day when we actually consider using `clippy-action` :) And it seems to me like you don’t disagree with the idea, if restricted to splitting the nextest step into two only?

Reply:
> That being said, this PR is only about nextest splitting, so I’d argue that the discussion about clippy lints could be postponed for the day when we actually consider using clippy-action :) And it seems to me like you don’t disagree with the idea, if restricted to splitting the nextest step into two only?

I mean… whatever plan we come up and start following should scale to whatever future ambitions we may have, which is why I think discussing clippy and lychee is also important… In particular one other thing I would (eventually, not necessarily in this PR) love for the setup to have is *automatic* determination of which `just` commands to run for the CI.

This could be something similar to the `wasmtime`’s [`determine` job](https://github.com/near/wasmtime/blob/main/.github/workflows/main.yml#L125) which generates a job matrix for the successive jobs. In our case we’d parse something along the lines of `just --summary` and generate a job matrix consisting of all the individual subcommands to run across different builders.

One thing I already like about `just` setup is that it also allows for maintaining the different flag combinations for different OSes (so we can move `--ignore integration-tests` on macos down into the justfile.)

Reply:
> In particular one other thing I would (eventually, not necessarily in this PR) love for the setup to have is automatic determination of which just commands to run for the CI.

While I’m philosophically of the same opinion, at the same time I’m wondering whether the time to implement and the effort to maintain such a determination are financially worth the savings in compute time 😅 but it’d be really nice to have that!